### DISCUSSION TITLE
Duplicate messages on overlapping subscriptions

### DISCUSSION THREAD

**Discussion by severn-everett** [emqx/emqx#11303] [Q&A] [answered]
Is there a specific configuration setting that needs to be enabled/disabled so that two subscriptions to overlapping topics don't cause EMQX to generate duplicate messages? Take, for example, a subscription to `foo/+/a_topic` and `#` (to monitor all MQTT traffic during integration testing). Sending a message to the topic `foo/an_object/a_topic` is causing our instance of EMQX to generate one message each for the subscriptions to `foo/+/a_topic` and `#`, because both subscriptions match the topic of `foo/an_object/a_topic`. We did not have this issue when using Mosquitto, so I assume that this would be an issue with something that we've neglected to configure in EMQX.

* Client: Eclipse Paho 1.2.5
* MQTT Protocol: 3.1.1
* Server: EMQX 5.0.19

**Comment by severn-everett** [comment]
A small reproduction example using Kotlin and Paho:
```kotlin
fun main() {
    MqttClient("tcp://localhost:1883", "test_client", MemoryPersistence()).use { client ->
        val connOpts = MqttConnectOptions().apply {
            userName = "userName"
            password = "passWord".toCharArray()
        }

        client.setCallback(object : MqttCallback {
            override fun connectionLost(cause: Throwable?) {
                println("Connection lost")
            }

            override fun messageArrived(topic: String, message: MqttMessage) {
                println("Message arrived: $topic | ${message.id}")
            }

            override fun deliveryComplete(token: IMqttDeliveryToken) {
                // No-Op
            }
        })

        client.connect(connOpts)
        client.subscribe("testtopic/#")
        client.subscribe("testtopic/1")

        client.publish("testtopic/1", MqttMessage("TEST".toByteArray()))

        client.disconnect()
        Thread.sleep(1000L)
    }
}
```

**Comment by lafirest** [comment, marked as answer]
@severn-everett 
Thank you for your feedback, but sorry, I think both are correct behaviour.
According to the specification [MQTT-3.3.5-1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html), the server could deliver to all matched topics even if some of them are overlapped.
```
When Clients make subscriptions with Topic Filters that include wildcards, it is possible for a Client’s subscriptions to overlap so that a published message might match multiple filters. In this case the Server MUST deliver the message to the Client respecting the maximum QoS of all the matching subscriptions [MQTT-3.3.5-1]. In addition, the Server MAY deliver further copies of the message, one for each additional matching subscription and respecting the subscription’s QoS in each case. 
```
